sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIO.java

```diff
@@ -0,0 +1,748 @@
+
+    abstract Builder builder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setConnectionConfig(ConnectionConfig connectionConfig);
+      abstract Write build();
+    }
+
+    /**
+     * Define the {@link ConnectionConfig} to RabbitMQ server.
+     */
+    public Write withConnectionConfig(ConnectionConfig connectionConfig) {
+      checkArgument(connectionConfig != null, "connectionConfig can not be null");
+      return builder().setConnectionConfig(connectionConfig).build();
+    }
+
+    public Write withUri(String uri) {
+      checkArgument(uri != null, "uri can not be null");
+      return builder().setConnectionConfig(connectionConfig().withUri(uri)).build();
+    }
+
+    public Write withQueue(String queue) {
+      checkArgument(queue != null, "queue can not be null");
+      return builder().setConnectionConfig(connectionConfig().withQueue(queue)).build();
+    }
+
+    public Write withExchange(String exchange, String exchangeType) {
+      checkArgument(exchange != null, "exchange can not be null");
+      checkArgument(exchangeType.equals("direct")
+              || exchangeType.equals("topic")
+              || exchangeType.equals("headers")
+              || exchangeType.equals("fanout"),
+          "exchangeType should be direct, topic, headers, fanout");
+      return builder()
+          .setConnectionConfig(connectionConfig().withExchange(exchange, exchangeType)).build();
+    }
+
+    @Override
+    public PDone expand(PCollection<byte[]> input) {
+      input.apply(ParDo.of(new WriteFn(this)));
+      return PDone.in(input.getPipeline());
+    }
+
+    private static class WriteFn extends DoFn<byte[], Void> {
+
+      private final Write spec;
+
+      private transient Connection connection;
+      private transient Channel channel;
+
+      public WriteFn(Write spec) {
+        this.spec = spec;
+      }
+
+      @Setup
+      public void setup() throws Exception {
+        ConnectionFactory connectionFactory = spec.connectionConfig().createConnectionFactory();
+        connection = connectionFactory.newConnection();
+        channel = connection.createChannel();
+        if (channel == null) {
+          throw new IOException("No RabbitMQ channel available");
+        }
+        if (spec.connectionConfig().queue() != null) {
+          channel.queueDeclare(spec.connectionConfig().queue(), false, false, false, null);
+        } else if (spec.connectionConfig().exchange() != null) {
+          channel.exchangeDeclare(spec.connectionConfig().exchange(),
+              spec.connectionConfig().exchangeType());
+        }
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext processContext) throws IOException {
+        byte[] element = processContext.element();
+        if (spec.connectionConfig().queue() != null) {
+          channel.basicPublish("", spec.connectionConfig().queue(), null, element);
```

Comment:
This is a strange choice of only two publishing strategies: either publishing to a default exchange, or publishing to a given exchange with an empty routing key. Why not allow the full flexibility of RabbitMQ, where each message can have a body, a routing key and properties? (I suppose the exchange would be the same within a single RabbitMqIO.write())

There's also the "mandatory" flag; I suppose it can be supported later because handling it is complicated.

Reply:
Agree, I will add the routing key use case.